repo2docker/app.py

```diff
@@ -166,29 +183,24 @@ def _user_name_default(self):
 
     def fetch(self, url, ref, checkout_path):
         """Check out a repo using url and ref to the checkout_path location"""
-        try:
-            cmd = ['git', 'clone', '--recursive']
-            if not ref:
-                cmd.extend(['--depth', '1'])
-            cmd.extend([url, checkout_path])
-            for line in execute_cmd(cmd, capture=self.json_logs):
-                self.log.info(line, extra=dict(phase='fetching'))
-        except subprocess.CalledProcessError:
-            self.log.error('Failed to clone repository!',
-                           extra=dict(phase='failed'))
-            sys.exit(1)
-
-        if ref:
-            hash = check_ref(ref, checkout_path)
-            if hash is None:
-                self.log.error('Failed to check out ref %s', ref,
-                               extra=dict(phase='failed'))
-                sys.exit(1)
-            # If the hash is resolved above, we should be able to reset to it
-            for line in execute_cmd(['git', 'reset', '--hard', hash],
-                                    cwd=checkout_path,
-                                    capture=self.json_logs):
-                self.log.info(line, extra=dict(phase='fetching'))
+        # Pick a content provider based on URL
+        picked_content_provider = None
+        for CP in self.content_providers:
+            cp = CP()
```

Comment:
Just to clarify: Since content_providers is a Traitlets List, the individual list element (`CP`) which refers to a list item (i.e. contentproviders.Local) which begins with an uppercase letter can be used to instantiate an object of class `CP`. For example, if CP=contentproviders.Git, then `cp = Git()`  [traitlets doc note](https://traitlets.readthedocs.io/en/stable/config.html#python-configuration-files)

Reply:
I think that `self.content_providers` is a Traitlets list. I believe that it is the traitlets functionality that allows the instantiation in `cp = CP()`. To create an instance, one would call `Git(contentproviders)` or in this case using the shorthand `Git()`. Perhaps I'm missing how this is working under the hood, since I don't think one can instantiate a list element directly. 🤷‍♀️

Reply:
If traitlets help this is an accident, all I had in mind when I wrote this code is:
```python
In [1]: class A:
   ...:     def __init__(self):
   ...:         print("A all day")
   ...:

In [2]: class B:
   ...:     def __init__(self):
   ...:         print("Because")
   ...:

In [3]: classes = [A, B]

In [4]: classes[0]()
A all day
Out[4]: <__main__.A at 0x103dbde10>
```